tla-assignments/src/main/scala/at/forsyte/apalache/tla/assignments/SmtFreeSymbolicTransitionExtractor.scala

```diff
@@ -0,0 +1,174 @@
+      throwOnUseBeforeAssignment( s.unassignedVars )( star )
+      // if `name` is not yet assigned, this spot becomes an assignment
+      if ( s.unassignedVars.contains( name )) PartialState( s.unassignedVars - name, s.partialStrat :+ ex.ID )
+      else s
+
+    /** Base case, manual assignments */
+    case ex@OperEx( BmcOper.assign, OperEx( TlaActionOper.prime, NameEx( name ) ), star ) =>
+      // Similar to above, except manual assignments throw if spurious
+      throwOnUseBeforeAssignment( s.unassignedVars )( star )
+      if ( s.unassignedVars.contains( name )) PartialState( s.unassignedVars - name, s.partialStrat :+ ex.ID )
+      else {
+        val locString = getLocString( ex )
+        throw new AssignmentException( s"Manual assignment at $locString is spurious: $name is already assigned!" )
+      }
+
+    /** Conjunction */
+    case OperEx( TlaBoolOper.and, args@_* ) =>
+      // We sequentially update the partial state
+      args.foldLeft( s ) { getStrategyOptInternal(_, letInOperMap)(_) }
+    /** Disjunction */
+    case ex@OperEx( TlaBoolOper.or, args@_* ) =>
+      handleDisjunctionOrITE( s, letInOperMap,  args)
+
+    /** \E quantifier */
+    case OperEx( TlaBoolOper.exists, NameEx( _ ), star, subEx ) =>
+      // We need to check that star does not contain unassigned primes.
+      throwOnUseBeforeAssignment( s.unassignedVars )( star )
+      // if no exception is thrown, continue in setEx
+      getStrategyOptInternal( s, letInOperMap )( subEx )
+
+
+    /** ITE */
+    case OperEx( TlaControlOper.ifThenElse, star, thenAndElse@_* ) =>
+      // We need to check that star does not contain unassigned primes.
+      throwOnUseBeforeAssignment( s.unassignedVars )( star )
+      // The rest is equivalent to the disjunction case
+      handleDisjunctionOrITE( s, letInOperMap, thenAndElse)
+
+    /** Nullary LetIn */
+    case LetInEx( body, defs@_* ) =>
+      // Sanity check, all operators must be nullary and non-recursive
+      assert( defs.forall { d => d.formalParams.isEmpty && !d.isRecursive } )
+      // We memorize the let-in operators, to recall when we see an apply
+      val newMap = BodyMapFactory.makeFromDecls( defs, letInOperMap)
+      // finally, we check the let-in body as well
+      getStrategyOptInternal( s, newMap )( body )
+
+    /** Nullary apply */
+    case OperEx( TlaOper.apply, NameEx(operName) ) =>
+      // Apply may appear in higher order operators, so it might not be possible to pre-analyze
+      val knownBodyOpt = letInOperMap.get( operName ) map { _.body }
+      val newStateOpt = knownBodyOpt map { getStrategyOptInternal(s, letInOperMap)(_) }
+
+      newStateOpt.getOrElse( s )
+
+    /** Misc. operator */
+    case OperEx( _, args@_* ) =>
+      // For other operators, make sure they don't use unassigned variables ...
+      args foreach throwOnUseBeforeAssignment( s.unassignedVars )
+      // ... but don't update assignments
+      s
```

Comment:
It is easy to miss `s` here. Actually, it is quite a short parameter name for such a long function. Shall we call it `state` or `partialState`?